lib/jsonrpc_client.ml

```diff
@@ -17,49 +17,101 @@
+				| n ->
+					if n > max_bytes
+					then
+					begin
+						debug "exceeding maximum read limit %d, clear buffer" !json_rpc_max_len;
+						Buffer.clear buf; (* otherwise might be parse error *)
+						Buffer.contents buf
+					end
+					else
+					begin
+						let used_time = get_used_time c in
+						let remain_time = Int64.sub max_time used_time in
+						(* here we want to make an exception in case the time is due but there are still some bytes to read, 
+						 * in such case we would like to expire a little to complete the read if possible. *)
+						let remain_time' = if remain_time < 0L then 0L else remain_time in
+						Buffer.add_subbytes buf bytes 0 n;
+						inner remain_time' (max_bytes - n)
+					end
+				| exception Unix.Unix_error(err,_,_) when err = Unix.EAGAIN || err = Unix.EWOULDBLOCK ->
+					let used_time = get_used_time c in
+					let remain_time = Int64.sub max_time used_time in
+					if remain_time < 0L then raise Timeout
+					else inner remain_time max_bytes
+		end
+		else
+		begin
+			let used_time = get_used_time c in
+			let remain_time = Int64.sub max_time used_time in
+			if remain_time < 0L then raise Timeout
+			else inner remain_time max_bytes
+		end
 	in
-	get ();
-	Buffer.contents buf
-
-let receive fin =
-	let obj = input_json_object fin in
-	debug "Response: %s" obj;
-	Jsonrpc.response_of_string obj
+	inner timeout !json_rpc_max_len
 
-let with_connection sockaddr f =
-	let fin, fout = Unix.open_connection sockaddr in
-	debug "Connected.";
-	let result = f fin fout in
-	Unix.shutdown_connection fin;
-	close_in fin;
-	debug "Shut down.";
-	result
+(* Write as many bytes to a file descriptor as possible from data before a given clock time. *)
+(* Raises Timeout exception if the number of bytes written is less than the specified length. *)
+(* Writes into the file descriptor at the current cursor position. *)
+let timeout_write filedesc length data response_time =
+	let rec inner_write filedesc data offset length remain_time =
+		let c = Mtime_clock.counter () in
+		let get_used_time counter = Mtime.Span.to_uint64_ns (Mtime_clock.count counter) in
+		let (_, ready_to_write, _) = Unix.select [] [filedesc] [] (to_s remain_time) in 
+		if List.mem filedesc ready_to_write then 
+		begin
+			let bytes_written = 
+				(try Unix.single_write filedesc data offset length with 
+				| Unix.Unix_error(Unix.EAGAIN,_,_)
+				| Unix.Unix_error(Unix.EWOULDBLOCK,_,_) -> 0)
+			in
+			let new_offset = offset + bytes_written in
+			let new_length = length - bytes_written in
+			let used_time = get_used_time c in
+			let new_remain_time = Int64.sub remain_time used_time in
+			if new_length = 0 then ()
+			else
+			if new_remain_time < 0L then raise Timeout
+			else inner_write filedesc data new_offset new_length new_remain_time
+		end
+		else
+		begin
+			let used_time = get_used_time c in
+			let new_remain_time = Int64.sub remain_time used_time in
+			if new_remain_time < 0L then raise Timeout
```

Comment:
When raising timeout, it's better to include the information of the size of the bytes we have read and the size of the bytes we expect to read.